File: app/src/main/java/com/devusercode/upchat/LoginActivity.java
package com.devusercode.upchat;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;

import com.devusercode.upchat.utils.SketchwareUtil;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.material.appbar.AppBarLayout;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException;
import com.google.firebase.auth.FirebaseAuthInvalidUserException;

public class LoginActivity extends AppCompatActivity {
    private final String TAG = this.getClass().getSimpleName();
    private final FirebaseAuth auth = FirebaseAuth.getInstance();

    private LinearLayout linear1;
    private EditText email_edit;
    private EditText password_edit;
    private TextView forgot_password_text;

    private OnCompleteListener<AuthResult> auth_create_user_listener;
    private OnCompleteListener<AuthResult> auth_sign_in_listener;
    private OnCompleteListener<Void> auth_reset_password_listener;
    private OnCompleteListener<Void> auth_updateEmailListener;
    private OnCompleteListener<Void> auth_updatePasswordListener;
    private OnCompleteListener<Void> auth_emailVerificationSentListener;
    private OnCompleteListener<Void> auth_deleteUserListener;
    private OnCompleteListener<Void> auth_updateProfileListener;
    private OnCompleteListener<AuthResult> auth_phoneAuthListener;
    private OnCompleteListener<AuthResult> auth_googleSignInListener;

    private final Intent intent = new Intent();

    @Override
    protected void onCreate(Bundle _savedInstanceState) {
        super.onCreate(_savedInstanceState);
        setContentView(R.layout.activity_login);
        initialize(_savedInstanceState);
        FirebaseApp.initializeApp(this);
        initializeLogic();
    }

    @Override
    public void startActivity(Intent intent) {
        super.startActivity(intent);
        overridePendingTransition(R.anim.right_in, R.anim.left_out);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.left_in, R.anim.right_out);
    }

    private void initialize(Bundle savedInstanceState) {
        CoordinatorLayout coordinator = findViewById(R.id._coordinator);
        AppBarLayout app_bar = findViewById(R.id.app_bar);
        Toolbar toolbar = findViewById(R.id.toolbar);

        setSupportActionBar(toolbar);
        ((TextView) toolbar.findViewById(R.id.toolbar_title)).setText("Login");

        linear1 = findViewById(R.id.linear1);
        email_edit = findViewById(R.id.email_edit);
        password_edit = findViewById(R.id.password_edit);
        forgot_password_text = findViewById(R.id.forgot_password_text);

        Button login_button = findViewById(R.id.login_button);
        Button register_button = findViewById(R.id.register_button);

        forgot_password_text.setOnClickListener(view -> {
            if (email_edit.getText().toString().isEmpty() || password_edit.getText().toString().isEmpty()) {
                return;
            }
            auth.sendPasswordResetEmail(email_edit.getText().toString()).addOnCompleteListener(auth_reset_password_listener);
        });

        login_button.setOnClickListener(view -> {
            if (email_edit.getText().toString().isEmpty() || password_edit.getText().toString().isEmpty()) {
                return;
            }
            auth.signInWithEmailAndPassword(email_edit.getText().toString(), password_edit.getText().toString())
                    .addOnCompleteListener(auth_sign_in_listener);
        });

        register_button.setOnClickListener(view -> {
            intent.setClass(getApplicationContext(), RegisterActivity.class);
            startActivity(intent);
        });

        /* *** LISTENERS *** */

        auth_emailVerificationSentListener = task -> {
            final boolean success = task.isSuccessful();
            final String errorMessage = task.getException() != null ? task.getException().getMessage() : "";
        };

        auth_googleSignInListener = task -> {
            final boolean success = task.isSuccessful();
            final String errorMessage = task.getException() != null ? task.getException().getMessage() : "";
        };

        auth_sign_in_listener = task -> {
            if (task.isSuccessful()) {
                intent.setClass(getApplicationContext(), MyProfileActivity.class);
                startActivity(intent);
                finish();
            } else {
                if (task.getException() instanceof FirebaseAuthInvalidUserException) {
                    SketchwareUtil.showMessage(getApplicationContext(), "User not found!");
                }

                if (task.getException() instanceof FirebaseAuthInvalidCredentialsException) {
                    SketchwareUtil.showMessage(getApplicationContext(), "Invalid credentials!");
                } else {
                    Log.e(TAG, task.getException().getClass().getName());
                    Log.e(TAG, task.getException().getMessage());
                }
            }
        };

        auth_reset_password_listener = task -> {
            final boolean success = task.isSuccessful();

            if (success) {
                SketchwareUtil.showMessage(getApplicationContext(), "Password reset email successfully sent");
            } else {
                Log.e(TAG, task.getException().getMessage());
                SketchwareUtil.showMessage(getApplicationContext(), task.getException().getMessage());
            }
        };
    }

    private void initializeLogic() {
    }

}

File: app/src/main/java/com/devusercode/upchat/LoginActivity.kt
package com.devusercode.upchat

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.CheckBox
import android.widget.EditText
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.devusercode.upchat.utils.StorageController
import com.devusercode.upchat.utils.applyActivityCloseAnimation
import com.devusercode.upchat.utils.applyActivityOpenAnimation
import com.devusercode.upchat.utils.setComposeContent
import com.devusercode.upchat.utils.Util
import com.devusercode.upchat.utils.applyActivityCloseAnimation
import com.devusercode.upchat.utils.applyActivityOpenAnimation
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.FirebaseApp
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException
import com.google.firebase.auth.FirebaseAuthInvalidUserException

@RequiresApi(Build.VERSION_CODES.O)
class LoginActivity : AppCompatActivity() {
    private val TAG = this.javaClass.simpleName
    private val auth = FirebaseAuth.getInstance()

    private lateinit var contentView: View
    private lateinit var emailEdit: EditText
    private lateinit var passwordEdit: EditText
    private lateinit var saveLoginInfoCheckbox: CheckBox
    private lateinit var forgotPasswordText: TextView

    // private lateinit var authCreateUserListener: OnCompleteListener<AuthResult>
    private lateinit var authSignInListener: OnCompleteListener<AuthResult>
    private lateinit var authResetPasswordListener: OnCompleteListener<Void>
    // private lateinit var authUpdateEmailListener: OnCompleteListener<Void>
    // private lateinit var authUpdatePasswordListener: OnCompleteListener<Void>
    // private lateinit var authEmailVerificationSentListener: OnCompleteListener<Void>
    // private lateinit var authDeleteUserListener: OnCompleteListener<Void>
    // private lateinit var authUpdateProfileListener: OnCompleteListener<Void>
    // private lateinit var authPhoneAuthListener: OnCompleteListener<AuthResult>
    // private lateinit var authGoogleSignInListener: OnCompleteListener<AuthResult>

    private val intent = Intent()
    private lateinit var storageController: StorageController

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        contentView = setComposeContent(R.layout.activity_login)

        storageController = StorageController.getInstance(this)!!

        initialize()
        initializeLogic()
        FirebaseApp.initializeApp(this)
    }

    override fun startActivity(intent: Intent) {
        super.startActivity(intent)
        applyActivityOpenAnimation(R.anim.right_in, R.anim.left_out)
    }

    override fun finish() {
        super.finish()
        applyActivityCloseAnimation(R.anim.left_in, R.anim.right_out)
    }

    private fun initialize() {
        emailEdit = contentView.findViewById(R.id.email_edit)
        passwordEdit = contentView.findViewById(R.id.password_edit)
        saveLoginInfoCheckbox = contentView.findViewById(R.id.save_login_info_checkbox)
        forgotPasswordText = contentView.findViewById(R.id.forgot_password_text)

        val loginButton: Button = contentView.findViewById(R.id.login_button)
        val registerButton: Button = contentView.findViewById(R.id.register_button)

        saveLoginInfoCheckbox.isChecked = storageController.getBool("save_login_info")

        forgotPasswordText.setOnClickListener {
            if (emailEdit.text.toString().isEmpty()) {
                emailEdit.error = "You need to fill this field!"
                return@setOnClickListener
            }

            auth
                .sendPasswordResetEmail(emailEdit.text.toString())
                .addOnCompleteListener(authResetPasswordListener)
        }

        loginButton.setOnClickListener {
            if (emailEdit.text.toString().isEmpty() || passwordEdit.text.toString().isEmpty()) {
                return@setOnClickListener
            }

            auth
                .signInWithEmailAndPassword(emailEdit.text.toString(), passwordEdit.text.toString())
                .addOnCompleteListener(authSignInListener)
        }

        registerButton.setOnClickListener {
            intent.setClass(applicationContext, RegisterActivity::class.java)
            startActivity(intent)
        }

        // *** LISTENERS ***

        /*
        authEmailVerificationSentListener = OnCompleteListener { task ->
            val success = task.isSuccessful
            val errorMessage = task.exception?.message ?: ""
        }
        authGoogleSignInListener = OnCompleteListener { task ->
            val success = task.isSuccessful
            val errorMessage = task.exception?.message ?: ""
        }
         */

        authSignInListener =
            OnCompleteListener { task ->
                if (task.isSuccessful) {
                    val saveLoginInfo = saveLoginInfoCheckbox.isChecked
                    storageController["save_login_info"] = saveLoginInfo

                    if (saveLoginInfo) {
                        storageController["email"] = emailEdit.text.toString()
                        storageController["password"] = passwordEdit.text.toString()
                    }

                    intent.setClass(applicationContext, HomeActivity::class.java)
                    startActivity(intent)
                    finish()
                } else {
                    if (task.exception is FirebaseAuthInvalidUserException) {
                        Util.showMessage(applicationContext, "User not found!")
                    }

                    if (task.exception is FirebaseAuthInvalidCredentialsException) {
                        Util.showMessage(applicationContext, "Invalid credentials!")
                    }

                    Log.e(TAG, task.exception?.javaClass?.name ?: "")
                    Log.e(TAG, task.exception?.message ?: "")
                }
            }

        authResetPasswordListener =
            OnCompleteListener { task ->
                val success = task.isSuccessful

                if (success) {
                    Util.showMessage(applicationContext, "Password reset email sent")
                } else {
                    Log.e(TAG, task.exception?.message ?: "")
                    Util.showMessage(applicationContext, task.exception?.message ?: "")
                }
            }
    }

    private fun initializeLogic() {
        val saveLoginInfo = storageController.getBool("save_login_info")

        if (saveLoginInfo) {
            val email = storageController.getString("email")
            val password = storageController.getString("password")

            if (email != null && password != null) {
                Log.d(TAG, "login by saved info")
                auth
                    .signInWithEmailAndPassword(email, password)
                    .addOnCompleteListener(authSignInListener)
            } else {
                Log.d(TAG, "email or password is not set!")
            }
        }
    }
}
